File: examples/dotnet/BasicUsage.cs
// How to use skaidb from C# — an ADO.NET provider (DbConnection/DbCommand).
//
//   dotnet run --project examples/dotnet -- [host] [port] [user] [password]

using System;
using System.Globalization;
using Skaidb;

string host = args.Length > 0 ? args[0] : "localhost";
int port = args.Length > 1 ? int.Parse(args[1], CultureInfo.InvariantCulture) : 7000;
string user = args.Length > 2 ? args[2] : "anonymous";
string password = args.Length > 3 ? args[3] : "";

string connString = $"Host={host};Port={port};User={user};Password={password};Consistency=Quorum";

using var conn = new SkaidbConnection(connString);
conn.Open();

// --- DDL ---
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "DROP TABLE IF EXISTS people";
    cmd.ExecuteNonQuery();
}
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "CREATE TABLE people (PRIMARY KEY (id))";
    cmd.ExecuteNonQuery();
}

// --- Batch insert with bound parameters (`?`) ---
var seed = new (long Id, string Name, long Age)[]
{
    (1, "Ada", 36),
    (2, "Linus", 54),
    (3, "Margaret", 80),
};
foreach (var (id, name, age) in seed)
{
    using var cmd = conn.CreateCommand();
    cmd.CommandText = "INSERT INTO people (id, name, age) VALUES (?, ?, ?)";
    cmd.Parameters.Add(id);
    cmd.Parameters.Add(name);
    cmd.Parameters.Add(age);
    cmd.ExecuteNonQuery();
}

// --- Query ---
Console.WriteLine("age > 40:");
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "SELECT id, name, age FROM people WHERE age > ? ORDER BY id";
    cmd.Parameters.Add(40L);
    using var reader = cmd.ExecuteReader();
    while (reader.Read())
    {
        Console.WriteLine($"  {reader.GetInt64(0)}  {reader.GetString(1)}  {reader.GetInt64(2)}");
    }
}

// --- Update ---
int updated;
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "UPDATE people SET age = ? WHERE id = ?";
    cmd.Parameters.Add(37L);
    cmd.Parameters.Add(1L);
    updated = cmd.ExecuteNonQuery();
}
Console.WriteLine($"updated {updated} row(s)");

// --- Point read by primary key ---
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "SELECT name, age FROM people WHERE id = ?";
    cmd.Parameters.Add(1L);
    using var reader = cmd.ExecuteReader();
    reader.Read();
    Console.WriteLine($"id=1: {reader.GetString(0)} {reader.GetInt64(1)}");
}

// --- Error handling ---
try
{
    using var cmd = conn.CreateCommand();
    cmd.CommandText = "SELECT * FROM does_not_exist";
    cmd.ExecuteNonQuery();
}
catch (SkaidbException e)
{
    Console.WriteLine($"expected error: {e.Message}");
}

// --- Delete + cleanup ---
int deleted;
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "DELETE FROM people WHERE id = ?";
    cmd.Parameters.Add(2L);
    deleted = cmd.ExecuteNonQuery();
}
Console.WriteLine($"deleted {deleted} row(s)");
using (var cmd = conn.CreateCommand())
{
    cmd.CommandText = "DROP TABLE people";
    cmd.ExecuteNonQuery();
}

File: examples/python/basic_usage.py
"""How to use skaidb from Python — DB-API 2.0 driver (like sqlite3 / psycopg2).

    python3 basic_usage.py [host] [port] [user] [password]

Uses the driver at ../../drivers/python. In a real project, install it with
`pip install -e drivers/python` (or vendor it) instead of the sys.path hack
below, which exists only so this example runs straight out of the repo.
"""
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "drivers" / "python"))
import skaidb  # noqa: E402

host = sys.argv[1] if len(sys.argv) > 1 else "localhost"
port = int(sys.argv[2]) if len(sys.argv) > 2 else 7000
user = sys.argv[3] if len(sys.argv) > 3 else "anonymous"
password = sys.argv[4] if len(sys.argv) > 4 else ""

with skaidb.connect(host=host, port=port, user=user, password=password) as conn:
    cur = conn.cursor()

    # --- DDL ---
    cur.execute("DROP TABLE IF EXISTS people")
    cur.execute("CREATE TABLE people (PRIMARY KEY (id))")

    # --- Batch insert with bound parameters (`?`) ---
    # Parameters are quoted client-side (see drivers/PROTOCOL.md §5); never
    # build SQL by string-formatting user input directly.
    cur.executemany(
        "INSERT INTO people (id, name, age) VALUES (?, ?, ?)",
        [(1, "Ada", 36), (2, "Linus", 54), (3, "Margaret", 80)],
    )

    # --- Query ---
    cur.execute("SELECT id, name, age FROM people WHERE age > ? ORDER BY id", (40,))
    print("age > 40:")
    for row in cur:
        print(" ", row)

    # --- Update ---
    cur.execute("UPDATE people SET age = ? WHERE id = ?", (37, 1))
    print(f"updated {cur.rowcount} row(s)")

    # --- Point read by primary key (routed directly to the owning replica) ---
    cur.execute("SELECT name, age FROM people WHERE id = ?", (1,))
    print("id=1:", cur.fetchone())

    # --- Error handling ---
    try:
        cur.execute("SELECT * FROM does_not_exist")
    except skaidb.DatabaseError as e:
        print(f"expected error: {e}")

    # --- Delete + cleanup ---
    cur.execute("DELETE FROM people WHERE id = ?", (2,))
    print(f"deleted {cur.rowcount} row(s)")
    cur.execute("DROP TABLE people")
